### PR TITLE
feat: implemented bot challenge three

### PR DESCRIPTION
**Pull request by msacco8**

### CODE REVIEW

File: Forta-Bots-Challenge-3/src/network.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/network.ts]: Hey @msacco8, as per the challenge guidelines, it's important to avoid using a hardcoded endpoint. The bot should be able to run on all 3 chains independently (the chain is determined by the Forta node running the bot), with the _respective logic executed on each side_ (L1 - L2).

I recommend finding another way to communicate between the two sides and of course if you get stuck feel free to ping one of us

File: Forta-Bots-Challenge-3/src/agent.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/agent.ts]: Although I recommend changing the logic to not use hardcoded endpoints, if you were to keep it this way, this should use the L2's latest blocknumber (if I run the bot on Ethereum, the Ethereum's blockNumber is passed which results in false positive alerts)

**Review by Vxatz — CHANGES_REQUESTED** [review]

File: Forta-Bots-Challenge-3/package.json
Comment (RCantu92) [inline Forta-Bots-Challenge-3/package.json]:
```suggestion
  "description": "Detects when the MakerDAO Bridge invariant is violated on Arbitrum and/or Optimism",
  "longDescription": "This bot checks the DAI balance of L1 escrow contracts and the total supply of L2 DAI contracts in a block to emit an alert when the bridge invariant is violated on Optimism and/or Arbitrum",
```
Small fix, as the alerts would be for either or both being violated. Not only when they both are.

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md]:
```suggestion
This bot detects when the MakerDAO bridge invariant is violated on Arbitrum and/or Optimism
```
Same as `package.json`

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md]: I think there could be a more descriptive name. I'll let you decide, but one that hints at the logic of the bot.

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md:19]: I would say, given the nature of what this bot is monitoring, it would be of both higher severity and type.

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md:37]: If you intended to include a block, it seems to have been accidentally omitted.

File: Forta-Bots-Challenge-3/src/constants.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/constants.ts]: Since they are the same value, they could be combined to a single constant. But keep a clear name like you have. Perhaps something like `OP_&_ARB_DAI_ADDRESS`? Or `L2_DAI_ADDRESS` with a comment above above saying `// Optimism & Arbitrum DAI address`? Just brainstorming.

File: Forta-Bots-Challenge-3/src/network.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/network.ts]: To piggyback off of @Vxatz's comment, if implemented correctly, there should be no need to implement your own `NetworkManager` as the `GAM` has a `NetworkManager` you could use. [Here](https://github.com/NethermindEth/general-agents-module?tab=readme-ov-file#networkmanager), for quick access.

**Review by RCantu92 — COMMENTED** [review]
I will "+1" @Vxatz's comments. I am stopping a bit short of a completely thorough review as his comments were critical to implement the bot properly.

Very solid attempt, and it is understandable why you approached it the way you did.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: forta config is included in the `.dockerignore` so it won't be available when the bot is deployed (and for a good reason, as it may contain sensitive data such as API keys or RPC URLs etc). You can hardcode it by declaring a `const BOT_ID = ""` in the constants file instead.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: Not sure why you chose `sendAlerts` over the usual `findings.push(..)`. I'm not very familiar with it, but it's a new SDK method meant to be used with those "new/V2/external" bots. (I tried running it and makes the bot crash for me)

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: An optimization suggestion would be to create an alert on this side only when the balance has changed from the previous alert

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: Perhaps this could be written more concisely using a combination of `first` and `blockSortDirection` in the get alerts (https://docs.forta.network/en/latest/sdk/#getalerts)

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: If I'm not mistaken, here we need the opposite condition for the invariant to be violated.

**Review by Vxatz — CHANGES_REQUESTED** [review]
Very close to completing the challenge! 💪

File: Forta-Bots-Challenge-3/src/agent.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.ts]: For clarity, you could declare a constant in `constants.ts` that is the Ethereum chainid. For example `ETH_CHAIN_ID = 1` and then use it here as `if (chainId === ETH_CHAIN_ID) { ... }` so the if check is a bit more self explanatory.

File: Forta-Bots-Challenge-3/src/finding.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/finding.ts]:
```suggestion
    description: `The amount of DAI held in Arbitrum's and Optimism's L1 escrow contracts`,
```
Very small fix

File: Forta-Bots-Challenge-3/src/finding.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/finding.ts]: I would recommend using an alert more specific to the task of the bot. Something making mention of Maker. You could take a quick look to our past bots for examples of what I mean.

Same comment regarding L27.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts]:
```suggestion
  if (chainId === OP_CHAIN_ID || chainId === ARB_CHAIN_ID) {
    l2DaiContract = new Contract(l2DaiAddress, ERC20_ABI, provider);
  } else {
    throw new Error("You are running the bot in a non supported network");
  }
```
Suggestion to create constants for the Optimism and Arbitrum chain ids like for Ethereum for clarity. Additionally, since their DAI contract addresses are the same, you could combine the `if` checks.

Lastly, it seems the argument `l1DaiAddress` should be `l2DaiAddress`.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts]:
```suggestion
  l2DaiAddress: string
```
Small mistake.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts]: Reminder that if you change the alert id as I suggested, this would need to be updated as well.

To prevent having to update it in multiple places, you could also create a constant for this and use it in the findings creator file and here.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts:78]: I think this variable could be renamed for clarity. In the Forta docs, [here](https://docs.forta.network/en/latest/sdk/#getalerts), they used the name `results`. Something along those lines would be more clear.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts]:
```suggestion
      chainId === OP_CHAIN_ID
```
Inline with the other suggestions.

File: Forta-Bots-Challenge-3/src/finding.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/finding.ts]: Since the other finding creator function has a specific name, perhaps this one as well, for clarity. Something like `createInvariantViolationFindings()`.

File: Forta-Bots-Challenge-3/src/finding.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/finding.ts]: Better `alertId` name, though I would say having `NETHERMIND` in not necessary. Could replace with Maker.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/utils.ts]: Unless I missed it somewhere, `1: "Ethereum"` is not used and is unnecessary.

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md]: Suggestion to rename the alert id here and L24. Same reason as the `alertId`s in the finding creator file.

File: Forta-Bots-Challenge-3/src/agent.spec.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.spec.ts]: Is this necessary, since you are "overriding" `mockBlockEvent` _beforeEach_ test anyway? Couldn't you just declare it `mockBlockEvent = new TestBlockEvent()` here rather than L144?

I don't have much familiarity with `createBlockEvent` from the SDK, so I could be missing something.

File: Forta-Bots-Challenge-3/src/agent.spec.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.spec.ts]: Could add an additionally check here to confirm it did emit the alert expected.

File: Forta-Bots-Challenge-3/src/agent.spec.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.spec.ts:188]: Like Vasilis mentioned, the invariant being violated would mean there is more L2 DAI than in the L2's corresponding L1 Escrow. This test should be updated.

File: Forta-Bots-Challenge-3/src/agent.spec.ts
Comment (RCantu92) [reply Forta-Bots-Challenge-3/src/agent.spec.ts]: Applicable in the other tests as well.

File: Forta-Bots-Challenge-3/src/agent.spec.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.spec.ts]: As a good practice, we don't concat findings because it can lead you to believe something is correct if it isn't.

For example, if `optimismFindings` were to have zero findings, and `arbitrumFindings` were to have the two expected findings, in the end, you'd end up with both expected findings, but they wouldn't have been generated as you expected.

So instead, we would check "as we go". So check `optimismFindings` at its point of creation, and same for `arbitrumFindings`.

**Review by RCantu92 — COMMENTED** [review]
Good work on the pivot, Michael! Great to see the progress. I agree with @Vxatz in that you are fairly close. Also, I am impressed with your implementation for your test suite.

Lastly, I just want to point out that we don't typically _resolve_ a PR comment until a commit addressing that exact comment has been pushed.

For example, I saw some comments from Vasilis were resolved, but now when I take a look at the same chunk of code, it remains unchanged. So it was confusing for me as I expected them to have been changed since the comment has been resolved.

File: Forta-Bots-Challenge-3/README.md
Comment (RCantu92) [inline Forta-Bots-Challenge-3/README.md]: Better! Just our practice would be for them to be all caps. Applies to both.

File: Forta-Bots-Challenge-3/src/agent.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/agent.ts]: Would advise against using the type of `any`. Could take a look [here](https://github.com/forta-network/forta-bot-sdk/blob/master/sdk/alerts.api.ts#L54C79-L54C79) for the type.

File: Forta-Bots-Challenge-3/src/constants.ts
Comment (RCantu92) [inline Forta-Bots-Challenge-3/src/constants.ts]: Same as in the README about the usage of all caps.

**Review by RCantu92 — COMMENTED** [review]
Good job, Michael! Left some very minor comments, but consider it approved on my end. I'll let Vasilis be the one to provide the final approval before merging.

File: Forta-Bots-Challenge-3/src/utils.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/utils.ts]: This doesn't work as expected with BigNumbers. You either need to convert them both to strings or use the BigNumber's `.eq()` operator.
```suggestion
    !prevEscrowBalance["optimismEscrowBalance"].eq(optimismEscrowBalance) ||
    !prevEscrowBalance["arbitrumEscrowBalance"] .eq(arbitrumEscrowBalance)
```

File: Forta-Bots-Challenge-3/src/agent.ts
Comment (Vxatz) [inline Forta-Bots-Challenge-3/src/agent.ts:21]: Please note that for values that can be initialized only once, there's a special `Initialize` block that should be used instead (which should be exported just like `handleBlock`)

**Review by Vxatz — APPROVED** [review]
Thanks for quickly taking care of the `sendAlerts` issue!

I left two comments to consider for future implementations, but feel free to merge! Good job!

**Pull request merged**
